src/lib/core/style/focus-classes.ts

```diff
@@ -111,10 +202,24 @@ export class FocusOriginMonitor {
         (focusTarget == this._lastTouchTarget || focusTarget.contains(this._lastTouchTarget));
   }
 
-  /** Handles focus events on a registered element. */
-  private _onFocus(event: Event, element: Element, renderer: Renderer,
-                   subject: Subject<FocusOrigin>) {
-    // If we couldn't detect a cause for the focus event, it's due to one of two reasons:
+  /**
+   * Handles focus events on a registered element.
+   * @param event The focus event.
+   * @param element The monitored element.
+   */
+  private _onFocus(event: FocusEvent, element: Element) {
+    // NOTE(mmalerba): We currently set the classes based on the focus origin of the most recent
+    // focus event affecting the monitored element. If we want to use the origin of the first event
+    // instead we should check for the cdk-focused class here and return if the element already has
+    // it. (This only matters for elements that have includesChildren = true).
+
+    // If we are not counting child-element-focus as focused, make sure that the event target is the
+    // monitored element itself.
+    if (!this._elementInfo.get(element).checkChildren && element != event.target) {
```

Comment:
I think that we should use `===` as often as possible. Also in the `setClasses` method.